### PR TITLE
Update BigQuery values formatting 

### PR DESCRIPTION
**Pull request by blowmage**
This PR focuses on the formatting of values returned from data listing and queries. Several fixes have been made to repeated scalar values, nested record values, repeated record values, and nested repeated record values. All returned values should now matching the format used by query parameters.

This means that dates and timestamps will be returned as Date and Time objects, instead of strings. And when using Standard SQL records will be returned as a Hash, and repeated values as an Array. These objects can also be used while inserting.

### CODE REVIEW

**Comment by blowmage** [comment]
Here is an example of some of the functionality that this PR enables:

```ruby
require "google/cloud/bigquery"

bigquery = Google::Cloud::Bigquery.new
db = bigquery.dataset "mine"

# nested and repeated records are flattened by default
db.query "SELECT * FROM stuff"
rows == [{"name"=>"mike", "foo_bar"=>"hey", "foo_baz_bif"=>1}, {"name"=>"mike", "foo_bar"=>"world", "foo_baz_bif"=>2}] #=> true

# use standard sql to return properly formatted values
db.query "SELECT * FROM stuff", standard_sql: true
rows == [{"name"=>"mike", "foo"=>[{"bar"=>"hey", "baz"=>{"bif"=>1}}, {"bar"=>"world", "baz"=>{"bif"=>2}}]}] #=> true

```

File: google-cloud-bigquery/acceptance/bigquery/advanced_test.rb [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    rows = bigquery.query "SELECT * FROM #{dataset_id}.#{table_id} WHERE id = ?", params: [2]
+
+    rows.class.must_equal Google::Cloud::Bigquery::QueryData
+    rows.count.must_equal 1
+    row = rows.first
+
+    assert_rows_equal rows.first, example_table_rows[1]
+  end
+
+  it "queries repeated scalars in legacy mode" do
```
Comment (landrito) [inline google-cloud-bigquery/acceptance/bigquery/advanced_test.rb]: Can you point me to some documentation explaining legacy?

File: google-cloud-bigquery/acceptance/bigquery/advanced_test.rb [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+  let(:dataset) { bigquery.dataset(dataset_id) || bigquery.create_dataset(dataset_id) }
+  let(:table_id) { "examples_table" }
+  let(:table) { @table }
+
+  before do
+    @table = get_or_create_example_table dataset, table_id
+  end
+
+  it "loads and returns the properly formatted data" do
+    table.data.sort_by { |r| r["id"] }.zip(example_table_rows) do |returned_row, example_row|
```
Comment (blowmage) [inline google-cloud-bigquery/acceptance/bigquery/advanced_test.rb]: It yields an element from the left list and an element from the right list. Very handy

File: google-cloud-bigquery/acceptance/bigquery/advanced_test.rb [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
...AAACqaXHeAAAAAXNSR0IArs4c6QAAABxpRE9UAAAAAgAAAAAAAAAgAAAAKAAAACAAAAAgAAABxj2CfowAAAGSSURBVHgB7Jc9TsNAEIX3JDkCPUV6KlpKFHEGlD4nyA04ACUXQKTgCEipUnKGNEbP0otentayicZ24SlWs7tjO/N9u/5J2b2+NUtuZcnwYE8BuQPyGZAPwXwLLPk5kG+BJa9+fgfkh1B+CeancL4F8i2Q/wWm/S/w+XFoTseftn0dvhu0OXfhpM+AGvzcEiYVAFisPqE9zrETJhHAlXfg2lglMK9z0f3RBfB+ZyRUV3x+erzsEIjjOBqc1xtNAIrvguybV3A9lkVHxlEE6GrrPb/ZvAySwlUnfCmlPQ+R8JCExvGtcRQBLFwj4FGkznX1VYDKPG/f2/MjwCksXACgdNUxJjwK9xwl4JihOwTFR0kIF+CABEPRnvsvPFctMoYKqAFSAFaMwB4pp3Y+bodIYL9WmIAaIOHxo7W8wiHvAjTvhUeNwwSgeAeAABbqOewC5hBdwFD4+9+7puzXV9fS6/b1wwT4tsaYAhwOOQdUQch5vgZCeAhAv3ZM31yYAAUgvApQQQ6n5w6FB/RVe1jdJOAPAAD//1eMQwoAAAGQSURBVO1UMU4DQQy8X9AgWopIUINEkS4VlJQo4gvwAV7AD3gEH4iSgidESpWSXyyZExP5lr0c7K5PsXBhec/2+jzjuWtent9CLdtu1mG5+gjz+WNr7IsY7eH+tvO+xfuqk4vz7CH91edFaF5v9nb6dBKm13edvrL+0Lk5lMzJkQDeJSkkgHF6mR8CHwMHCQR/NAQQGD0BAlwK4FCefQiefq+A2Vn29tG7igLAfmwcnJu/nJy3BMQkMN9HEPr8AL3bfBv7Bp+7/SoExMDjZwKEJwmyhnnmQIQEBIlz2x0iKoAvJkAC6TsTIH6MqRrEWUMSZF2zAwqT4Eu/e6pzFAIkmNSZ4OFT+VYBIIF//UqbJwnF/4DU0GwOn8r/JQYCpPGufEfJuZiA37ycQw/5uFeqPq4pfR6FADmkBCXjfWdZj3NfXW58dAJyB9W65wRoMWulryvAyqa05nQFaDFrpa8rwMqmtOZ0BWgxa6WvK8DKprTmdAVoMWulryvAyqa05nQFaDFrpa8rwMqmtOb89wr4AtQ4aPoL6yVpAAAAAElFTkSuQmCC"),
```
Comment (blowmage) [inline google-cloud-bigquery/acceptance/bigquery/advanced_test.rb]: I'd like to keep it in Ruby so I can control the values. This structure comes from Omar's integration test issue

File: google-cloud-bigquery/lib/google/cloud/bigquery/convert.rb [hunk 4]
```diff
@@ -183,21 +191,54 @@ def self.to_query_param value
...
+        # @private
+        def self.to_json_rows rows
+          rows.map do |row|
+            Hash[row.map { |k, v| [k.to_s, to_json_value(v)] }]
+          end
+        end
+        ##
+        # @private
+        def self.to_json_value value
+          if DateTime === value
```
Comment (blowmage) [inline google-cloud-bigquery/lib/google/cloud/bigquery/convert.rb]: Yeah, a case statement uses `===`. The IO support makes it hard to use a case statement though, so we have this syntax. We have similar code in Datastore and Spanner, for the same reason.

File: google-cloud-bigquery/acceptance/bigquery/advanced_test.rb (hunk 1 above)
Comment (blowmage) [inline google-cloud-bigquery/acceptance/bigquery/advanced_test.rb]: There isn't a super clear explanation online. We link to this URL in the docs:

https://cloud.google.com/bigquery/docs/reference/legacy-sql

File: google-cloud-bigquery/lib/google/cloud/bigquery/convert.rb [hunk 5]
```diff
@@ -183,21 +191,54 @@ def self.to_query_param value
...
+            value.strftime "%Y-%m-%d %H:%M:%S.%6N"
+          elsif Date === value
+            value.to_s
+          elsif ::Time === value
+            value.strftime("%Y-%m-%d %H:%M:%S.%6N%:z")
```
Comment (blowmage) [inline google-cloud-bigquery/lib/google/cloud/bigquery/convert.rb]: Fixed.

File: google-cloud-bigquery/acceptance/bigquery/advanced_test.rb (hunk 2 above)
Comment (landrito) [inline google-cloud-bigquery/acceptance/bigquery/advanced_test.rb]: I see interesting. 

Is it correct for me to group this together in my head as such?

```rb
table.data.sort_by( { |r| r["id"]} ).zip(example_table_rows) ...

Comment (blowmage) [inline google-cloud-bigquery/acceptance/bigquery/advanced_test.rb]: I don't understand your question. Sorry.

Comment (landrito) [inline google-cloud-bigquery/acceptance/bigquery/advanced_test.rb]: Ah was trying something strange in irb. Didn't pan out how I'd like. Disregard this.

**Pull request merged**